### ISSUE TITLE
Rpi3 Default UART Console

### ISSUE THREAD

**Issue by gagachang** [rawcoder/rpi3_secure_uart#1]
Hi, @rawcoder 

You really developed a nice tutorial for studying deeper into OP-TEE.
However, I have some questions about UART and OP-TEE's serial console.
Would you help me solve my doubt?
Any suggestion would be appreciated. Thank you.

From the documentation of OP-TEE, the Rpi3 platform uses UART as serial console to debug.
Any logs printed from OP-TEE kernel will be putted into the serial console, which is UART.
If you attached the GPS module to UART, would it conflict between GPS module and serial console?
And if the secure UART is used by GPS module, how could you debug the system?
Did you connect the Rpi to internet and use SSH?

**Comment by rawcoder** [comment]
Hi @gagachang, thanks for the kind words of appreciation.

> And if the secure UART is used by GPS module, how could you debug the system?

If I recall correctly I didn't use serial console or ssh.  Instead, I connected the RPi to a spare monitor and used that for debugging.

> Any logs printed from OP-TEE kernel will be putted into the serial console, which is UART.

I believe [`build.patch`](https://github.com/rawcoder/rpi3_secure_uart/blob/master/build.patch#L19) solves this by asking the kernel to disable serial console output.

>  Did you connect the Rpi to internet and use SSH?

You cannot use ssh as it is not available in the initial OP-TEE environment.  However, you can use it when the normal OS has booted.

Hope this helps!

**Comment by gagachang** [comment]
Hi, @rawcoder 

Thanks for your help. In fact, I am trying to develop a TA which receives data from secure UART too.
May I ask you more questions? Thank you.

>  I connected the RPi to a spare monitor and used that for debugging.

How about debugging the secure world? Did you port the driver of monitor into OP-TEE and register 
 it as console?

Because in Rpi3, OP-TEE registers the UART as serial console at boot time for debugging.
And your Pseudo TA calls the `console_getchar()` to get the GPS data from UART (GPS module).
But the OP-TEE's logs will also be output to this console ( From `DMSG()`, `EMSG()` ), which is now attaching the GPS module and we can't see the logs for debugging anymore.
I am stuck in this question. 😖

**Comment by rawcoder** [comment]
> How about debugging the secure world? Did you port the driver of monitor into OP-TEE and register it as another console?

No, I didn't port any driver.  I experimented with OP-TEE in QEMU before trying anything with real hardware.

> But the OP-TEE's logs will also be output to this console ( From DMSG(), EMSG() ), which is now attaching the GPS module and we can't see the logs for debugging anymore.

You are right, OP-TEE dumps its debug logs into serial console and you would have to disable OP-TEE debugging (there is an option for it).  This makes development a little harder but you can try things in QEMU first and if it works then you can move to real hardware.  As UART has a rather simple interface, it was relatively easier to work with.
